Use functors in NS BCs and allow for other-side evaluations

GiudGiud opened the pull request:
This will help connect stitched components. We ll need physics and the components refactor plus the abstracted physics components and components junctions to use this

GiudGiud committed:
Add the capability to evaluate functors from the other side of a DirichletBC
refs #25933

GiudGiud committed:
Make finite volume inlet BCs rely on functors instead of functions.
This allows to build connections between variables

GiudGiud committed:
Add more information to error message on flow BCs

GiudGiud committed:
Deprecate function for INSFVInletVelocityBC

GiudGiud committed:
Add test input for using a dirichlet BC from the other side

Add test for dirichlet BC on other side, refs #25933

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1857273/) on 882f7bf wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25934/site)

This comment will be updated on new commits.

GiudGiud commented on framework/src/fvbcs/FVFunctorDirichletBC.C:
```diff
@@ -20,18 +20,34 @@ FVFunctorDirichletBCTempl<is_ad>::validParams()
-  : FVDirichletBCBase(parameters), _functor(getFunctor<GenericReal<is_ad>>("functor"))
+  : FVDirichletBCBase(parameters),
+    _functor(getFunctor<GenericReal<is_ad>>("functor")),
+    _use_other_side(getParam<bool>("functor_defined_on_other_side"))
```
makes sense. on it

GiudGiud committed:
Address Alex's review: rename parameter to avoid confusion

lindsayad commented:
Noting my conceptual approval although those tests need to be fixed...

GiudGiud committed:
Add layers of ghosting if the functor values come from the
other side of the functor dirichlet FVBC
- seems up to 3 are needed for FV variable + two-term expansion

lindsayad reviewed (dismissed):
my button clicker is getting tired

GiudGiud commented:
you're not going to like this

GiudGiud commented:
I m not sure I understand why I need 3 layers of ghosting for this FV variable on the other side. I would expect two to be enough?
One to get the other-side-cell value and one to get its neighbors to get the gradient to go to the face ?

lindsayad commented:
> I m not sure I understand why I need 3 layers of ghosting for this FV variable on the other side. I would expect two to be enough?
> One to get the other-side-cell value and one to get its neighbors to get the gradient to go to the face ?

Yes I agree that two should be enough

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1860603/) on d98353d wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>191b9d</th>
      <th colspan=4>#25934 d98353</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.63%</td>
      <td>85.63%</td>
      <td>+0.00%</td>
      <td>87.50%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>92451</td>
      <td>92463</td>
      <td>+12</td>
      <td>14</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15509</td>
      <td>15511</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25934/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25934/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>191b9d</th>
      <th colspan=4>#25934 d98353</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.77%</td>
      <td>84.77%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12548</td>
      <td>12550</td>
      <td>+2</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2255</td>
      <td>2255</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25934/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25934/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25934/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 87.50% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented on test/tests/fvbcs/fv_functor_dirichlet/fv_other_side.i:
```diff
@@ -0,0 +1,103 @@
+    variable = u
+    functor = v
+    functor_only_defined_on_other_side = true
+    ghost_layers = 3
```
This doesn't make sense. We should figure it out

lindsayad replied:
I can look into it if you would prefer

GiudGiud replied:
do we need the point neighbor ghosting?
I was thinking we dont because the side values of the cell on the other side dont depend on the cells that only touch with a corner

GiudGiud replied:
but if we did that would explain the "3"

lindsayad replied:
We should need one side neighbor layer directly from the FV Dirichlet BC, and then if the functor is a block restricted variable we will need one additional side neighbor layer to evaluate its cell gradient, so two side neighbor layers. In theory two *point* neighbor layers would be overkill and one *point* neighbor layer would be insufficient

lindsayad replied:
Never mind, I can see how we would get 3. A non-interface face needs needs neighboring cell gradients for the un-corrected face gradient (1). The cell gradient calls the FV Dirichlet BC which triggers the neighbor face evaluation (2). The neighbor face evaluation for two term expansion calculates a cell gradient which requires neighbor face values (3)

lindsayad reviewed (approved):
It would be nice to see a CI result for this, but the code looks good to this human

moosebuild commented:
All [jobs](https://civet.inl.gov/event/151534/) on d98353d : invalidated by @GiudGiud

refresh ci run

moosebuild commented:
Job [Clang min](https://civet.inl.gov/job/1860597/) on d98353d : invalidated by @GiudGiud

refresh. looks unrelated, canont be too careful

moosebuild commented:
Job [Clang](https://civet.inl.gov/job/1860591/) on d98353d : invalidated by @GiudGiud

refresh, looks unrelated

moosebuild commented:
Job [Framework increased sparsity](https://civet.inl.gov/job/1860582/) on d98353d : invalidated by @GiudGiud

refresh, looks unrelated

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1860603/) on d98353d wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

GiudGiud commented:
All good on CI now. (Bison unrelated) Thanks for the review!
